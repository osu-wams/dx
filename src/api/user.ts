import axios from 'axios';
import useAPICall from './useAPICall';
import { useEffect, useState } from 'react';
import { defaultTheme } from '../theme/themes';

export const defaultCampus = 'C';

export const CLASSIFICATIONS = {
  firstYear: ['freshman', 'vet med-first year'],
  graduate: ['graduate', 'cascades partner grad course', 'e-campus graduate course']
};

export const CLASSIFICATION_AUDIENCES = {
  firstYear: 'First Year',
  international: 'International Student',
  graduate: 'Graduate Student'
};

export const CAMPUS_CODES = {
  bend: 'B',
  corvallis: 'C',
  ecampus: 'DSC'
};

export const AFFILIATIONS = {
  employee: 'employee',
  student: 'student'
};

export const GROUPS = {
  admin: 'admin',
  masquerade: 'masquerade'
};

export interface IUserClassificationAttributes {
  level: string;
  campus: string;
  campusCode: string;
  classification: string;
  isInternational: boolean;
}

export interface IUserClassification {
  attributes?: IUserClassificationAttributes;
  id?: string;
}

export interface IUser {
  email?: string;
  isCanvasOptIn?: boolean;
  isAdmin: boolean;
  groups: string[];
  classification: IUserClassification;
  audienceOverride: IUserAudienceOverride;
  theme: string;
  primaryAffiliation: string;
  primaryAffiliationOverride?: string;
}

export interface IUserState {
  data: IUser | null;
  error: boolean;
  loading: boolean;
  isCanvasOptIn?: boolean;
}

export interface IUserAudienceOverride {
  campusCode?: string;
  firstYear?: boolean;
  graduate?: boolean;
  international?: boolean;
}

export interface IUserSettings {
  audienceOverride?: IUserAudienceOverride;
  primaryAffiliationOverride?: string;
  theme?: string;
}

export const initialUser: IUser = {
  classification: {},
  audienceOverride: {},
  isAdmin: false,
  groups: [],
  theme: defaultTheme,
  primaryAffiliation: AFFILIATIONS.employee
};

const getUser = (): Promise<IUser> => axios.get('/api/user').then(res => res.data);

export const getClassification = (): Promise<IUserClassification> =>
  axios.get('/api/user/classification').then(res => res.data);

/**
 * Returns the audience override value or users classification in that order
 * of precedence.
 * @param user the user to inspect
 * @returns whether or not the override or user classification is true
 */
const isFirstYear = (user: IUser): boolean => {
  if (user?.audienceOverride?.firstYear !== undefined) {
    return user.audienceOverride.firstYear;
  }
  return (
    user.classification.attributes !== undefined &&
    user.classification.attributes.classification !== null &&
    CLASSIFICATIONS.firstYear.includes(
      user!.classification!.attributes!.classification.toLowerCase()
    )
  );
};

/**
 * Returns the audience override value or users classification in that order
 * of precedence.
 * @param user the user to inspect
 * @returns whether or not the override or user classification is true
 */
const isInternational = (user: IUser): boolean => {
  if (user.audienceOverride?.international !== undefined) {
    return user.audienceOverride.international;
  }
  return (
    user.classification.attributes !== undefined && user.classification.attributes.isInternational
  );
};

/**
 * Returns the audience override value or users classification in that order
 * of precedence.
 * @param user the user to inspect
 * @returns whether or not the override or user classification is true
 */
const isGraduate = (user: IUser): boolean => {
  if (user.audienceOverride?.graduate !== undefined) {
    return user.audienceOverride.graduate;
  }
  return (
    user.classification.attributes?.level !== undefined &&
    CLASSIFICATIONS.graduate.includes(user.classification.attributes!.level?.toLowerCase())
  );
};

/**
 * Returns whether or not the users current primaryAffiliation is one of the supplied affiliations.
 * This intends to check if the user is a student or an employee while giving the application the ability
 * to specify scenarios where the user is in a number of affiliations.
 * @param user the user to inspect
 * @param affiliations the affiliations to check if the user is associated with
 */
export const hasPrimaryAffiliation = (user: IUser, affiliations: string[]): boolean => {
  if (user.primaryAffiliationOverride) {
    return affiliations.includes(user.primaryAffiliationOverride);
  }
  return affiliations.includes(user.primaryAffiliation);
};

/**
 * Returns your primary affiliation or the affiliationOverride if one is present
 * @param user the user to inspect
 */
export const getAffiliation = (user: IUser): string => {
  return user.primaryAffiliationOverride ?? user.primaryAffiliation;
};

/**
 * This method returns a fully populated user settings theme and overrides taking into consideration
 * thier student classification as well as any potentially persisted overrides
 * @param user the user to inspect
 */
export const usersSettings = (user: IUser): IUserSettings => ({
  theme: user.theme,
  audienceOverride: {
    campusCode: usersCampus(user).campusCode,
    firstYear: isFirstYear(user),
    international: isInternational(user),
    graduate: isGraduate(user)
  },
  primaryAffiliationOverride: user.primaryAffiliationOverride
});

/**
 * Detect if the user setting matches the default, taking into consideration the student classification if it exists.
 * @param user the user to inspect
 * @param propertyName check the student classification property value
 * @param currentValue the value to consider when the student classification doesn't exist
 * @param defaultValue the default value for comparison
 */
export const settingIsDefault = (
  user: IUser,
  propertyName: string,
  currentValue: string,
  defaultValue: string
): boolean => {
  const {
    classification: { attributes }
  } = user;
  if (attributes) {
    return attributes[propertyName] === currentValue;
  } else {
    return currentValue === defaultValue;
  }
};

/**
 * Detect if the user setting is an override of the default, taking into consideration the student classification if it exists.
 * @param user the user to inspect
 * @param propertyName check the student classification property value
 * @param currentValue the value to consider when the student classification doesn't exist
 * @param defaultValue the default value for comparison
 */
export const settingIsOverridden = (
  user: IUser,
  propertyName: string,
  currentValue: boolean,
  defaultValue: boolean
): boolean => {
  const {
    classification: { attributes }
  } = user;
  if (attributes) {
    const { isInternational, classification, level } = attributes;
    switch (propertyName) {
      case 'international':
        if (isInternational && currentValue !== undefined) {
          return !currentValue;
        } else {
          return false;
        }
      case 'firstYear':
        if (
          CLASSIFICATIONS.firstYear.includes(classification?.toLowerCase()) &&
          currentValue !== undefined
        ) {
          return !currentValue;
        } else {
          return false;
        }
      case 'graduate':
        if (CLASSIFICATIONS.graduate.includes(level?.toLowerCase()) && currentValue !== undefined) {
          return !currentValue;
        } else {
          return false;
        }
      default:
        return false;
    }
  } else {
    return currentValue !== defaultValue;
  }
};

/**
 * Returns the audience override value or users classification in that order
 * of precedence.
 * @param user the user to inspect
 * @returns the campus name and campus code that the user is associated with
 */
export const usersCampus = (
  user: IUser
): { campusName: string | undefined; campusCode: string } => {
  const { campusCode } = user.classification?.attributes ?? {
    campusCode: defaultCampus
  };
  const campusCodeOverride = user.audienceOverride?.campusCode ?? defaultCampus;
  const selectedCampusCode = campusCodeOverride || campusCode;
  // Find the key name associated to the users campusCode to use for matching in the audiences
  // set for the announcement
  const campusName = Object.keys(CAMPUS_CODES)
    .map(k => k.toLowerCase())
    .find(key => CAMPUS_CODES[key].toLowerCase() === selectedCampusCode.toLowerCase());
  return { campusCode: selectedCampusCode, campusName };
};

/**
 * Detects if the user is associated with any of the audiences that are provided. An audience could be
 * the users campus code or the classifications (both of which take into account the user override settings)
 * @param user the user to inspect
 * @param item a list of audiences to detect
 */
export const hasAudience = (user: IUser, item: { audiences: string[] }): boolean => {
  const foundAudiences: string[] = [];
  const { audiences } = item;
  if (
    audiences?.length === 0 ||
    (user.classification?.attributes === undefined && user.audienceOverride === undefined)
  )
    return true;

  const { campusName, campusCode } = usersCampus(user) as {
    campusName: string | undefined;
    campusCode: string;
  };

  if (!campusName) {
    console.error(
      `Expected campus code ${campusCode} not found in configuration, this is an unexpected circumstance that needs to be repaired.`
    );
  } else {
    foundAudiences.push(campusName);
  }

  if (isGraduate(user)) foundAudiences.push(CLASSIFICATION_AUDIENCES.graduate);
  if (isFirstYear(user)) foundAudiences.push(CLASSIFICATION_AUDIENCES.firstYear);
  if (isInternational(user)) foundAudiences.push(CLASSIFICATION_AUDIENCES.international);

  // The user has a classification and the item has audiences specified, return if
  // this users campusCode exists in the audience list.
  return item.audiences.some(a =>
    foundAudiences.map(fa => fa.toLowerCase()).includes(a.toLowerCase())
  );
};

/**
 * Detect if the users classification indicates that they are part of the campus provided
 * @param user the user to inspect
 * @param code the campus code for comparison
 */
export const atCampus = (user: IUser, code: string): boolean => {
  const { campusCode } = usersCampus(user);
  return campusCode.toLowerCase() === code.toLowerCase();
};

/**
 * The primary hook to fetch the user session and set the user for access throughout the application, this
 * is intended to be set near the root level of the application and exposed by way of the UserContext.
 */
export const useUser = () => {
  const [user, setUser] = useState<IUserState>({
    data: initialUser,
    error: false,
    loading: true,
    isCanvasOptIn: false
  });
  const u = useAPICall<IUser>(getUser, undefined, data => data, initialUser, false);
  const classification = useAPICall<IUserClassification>(
    getClassification,
    undefined,
    data => data,
    {},
    true
  );

  useEffect(() => {
    setUser({
      data: { ...u.data, classification: { ...classification.data } },
      error: u.error,
      loading: u.loading,
      isCanvasOptIn: u.data.isCanvasOptIn
    });
  }, [u.data, u.error, u.loading, classification.data, classification.loading]);

  return {
    error: user!.error,
    data: user!.data,
    loading: user!.loading,
    isCanvasOptIn: user!.data!.isCanvasOptIn,
    setUser
  };
};

/**
 * Send the settings to the backend to be saved.
 * @param settings the settings to persist to the backend
 */
export const postSettings = (settings: IUserSettings): Promise<IUserSettings> =>
  axios
    .post('/api/user/settings', settings)
    .then(res => res.data)
    .catch(e => {
      console.error(e);
      throw e;
    });
