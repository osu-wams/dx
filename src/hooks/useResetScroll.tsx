import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

export const useResetScroll = () => {
  const navigate = useNavigate();
  const { state, pathname } = useLocation();

  const updateState = React.useCallback(() => {
    navigate(pathname, {
      // @ts-ignore spread object
      state: { ...state, scrolled: true },
      replace: true,
    });
    window.scrollTo(0, 0);
  }, [pathname, state, navigate]);

  React.useLayoutEffect(() => {
    // @ts-ignore type unknown
    if (!state) {
      updateState();
    }
  }, [state, updateState]);
};

export default useResetScroll;
