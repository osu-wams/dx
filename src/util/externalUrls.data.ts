const Url = {
  myDegrees: {
    main: 'https://mydegrees.oregonstate.edu/DashboardServlet'
  },
  canvas: {
    main: 'https://oregonstate.instructure.com',
    test: 'https://oregonstate.test.instructure.com'
  },
  campusMap: {
    main: 'https://map.oregonstate.edu/',
    building: 'https://map.oregonstate.edu/?building='
  },
  events: {
    academicCalendar: 'https://registrar.oregonstate.edu/osu-academic-calendar'
  },
  banner: {
    financialTransactions:
      'https://xe.ucsadm.oregonstate.edu:9990/ssomanager/c/SSB?pkg=OSU_ARWEB.P_ViewBillDate'
  },
  myosu: {
    main: 'https://myosu.oregonstate.edu/'
  },
  bill: {
    main: 'http://mybill.oregonstate.edu'
  },
  support: {
    main: 'https://oregonstate.qualtrics.com/jfe/form/SV_73vMvyQZBQx8aaN?type=support'
  },
  feedback: {
    main: 'https://oregonstate.qualtrics.com/jfe/form/SV_73vMvyQZBQx8aaN?type=feedback'
  },
  registrar: {
    academicStanding: 'https://registrar.oregonstate.edu/grades-honor-roll-academic-standing'
  },
  empcenter: {
    main: 'https://osu.workforcehosting.com/'
  },
  evals: {
    main: 'https://myosu.oregonstate.edu/web/myosu/employee?gpi=10190'
  },
  itSystemStatus: {
    main: 'https://status.is.oregonstate.edu'
  }
};

export default Url;
