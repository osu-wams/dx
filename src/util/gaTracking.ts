import ReactGA from 'react-ga';

/**
 * Event - Add custom tracking event.
 * @param {enum} category
 * @param {string} action
 * @param {string} label
 */
export const Event = (category: IComponents, action: string, label?: string) => {
  ReactGA.event({
    category,
    action,
    label,
  });
};

// To make it easier to filter, our components will be the categories in Google Analytics
export type IComponents =
  | 'footer'
  | 'beta'
  | 'beta-info'
  | 'header'
  | 'student-navigation-main'
  | 'employee-navigation-main'
  | 'student-mobile-nav'
  | 'footer-menu-nav'
  | 'employee-tools'
  | 'schedule-card'
  | 'resources-card'
  | 'favorite-resources-card'
  | 'favorite-resource'
  | 'trending-resources-card'
  | 'trending-resource'
  | 'anouncements'
  | 'events'
  | 'courses'
  | 'course'
  | 'planner-items'
  | 'meal-plans'
  | 'dx-event'
  | 'calendar-event'
  | 'academic-calendar'
  | 'program-of-study'
  | 'financial-transactions'
  | 'academic-overview'
  | 'account-balance'
  | 'resource'
  | '404'
  | 'resource-search'
  | 'resource-search-failed'
  | 'resource-category'
  | 'past-courses-search'
  | 'info-button'
  | 'canvas-auth'
  | 'badge'
  | 'it-system-status'
  | 'it-system-status-sticky'
  | 'training';
