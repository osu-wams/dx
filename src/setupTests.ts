import ReactGA from 'react-ga';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import * as cache from './util/cache';

ReactGA.initialize('UA-48705802-13', {
  testMode: true
});

export const mockGAEvent = jest.fn();

jest.mock('../src/util/gaTracking', () => ({
  Event: () => {
    return mockGAEvent();
  }
}));

mockGAEvent.mockResolvedValue(Promise.resolve(true));

beforeEach(() => {
  mockGAEvent.mockClear();
});

afterEach(() => {
  // don't output debug statements to console
  jest.spyOn(console, 'debug').mockImplementation(() => {});

  jest.clearAllMocks();
});

// required because of the overlay from Reakit
if (global.document) {
  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document
    }
  });
}

// Supress missing CSS warnings in tests from @reach ui components
jest.mock('@reach/utils', () => ({
  ...jest.requireActual('@reach/utils'),
  checkStyles: jest.fn()
}));

// Mock matchMedia for test env
const matchMedia = () => ({
  matches: false,
  addListener: () => {},
  removeListener: () => {}
});
window.matchMedia = window.matchMedia || matchMedia;

// Mock the location change method
window.location.assign = jest.fn();

// Mock sessionStorage interface
beforeEach(() => {
  Storage.prototype.clear = jest.fn();
  Storage.prototype.getItem = jest.fn();
  Storage.prototype.setItem = jest.fn();
  Storage.prototype.removeItem = jest.fn();
  cache.clear();
});
