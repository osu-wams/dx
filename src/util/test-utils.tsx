import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render as testingLibraryRender } from '@testing-library/react';
import { Context as ResponsiveContext } from 'react-responsive';
import { themesLookup, defaultTheme } from '../theme/themes';
import { User } from '@osu-wams/lib';
import { mobile, desktop } from 'src/util/useMediaQuery';
import { RecoilRoot } from 'recoil';
import { userState } from 'src/state/application';
import { rest } from 'msw';
import { server } from 'src/mocks/server';

// Helper method to change the mock responses by MSW
export const alterMock = (api: string, mock: any) => {
  server.use(
    rest.get(api, async (req, res, ctx) => {
      return res(ctx.json(mock));
    })
  );
};

const { mockUser } = User;
export const authUserAudienceOverride = mockUser.userAudienceOverride;

export const mockGradUser = {
  ...mockUser.userGraduate,
  refreshFavorites: jest.fn(),
};

export const authUserClassification = mockUser.userClassification;

export const mockEmployeeUser = {
  ...mockUser.userEmployee,
  refreshFavorites: jest.fn(),
};

export const mockStudentEmployeeUser = {
  ...mockUser.userStudentEmployee,
  refreshFavorites: jest.fn(),
};

// The default undegraduate user with canvas opted in
export const authUser = {
  ...mockUser.user,
  refreshFavorites: jest.fn(),
};

// The default undegraduate user with canvas opted in
export const mockAdminUser = {
  ...mockUser.userAdmin,
  refreshFavorites: jest.fn(),
};

const renderWithUserContext = (
  ui,
  { user = authUser, initialStates = new Array(), ...options } = {}
) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, user);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]} {...props} />
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const renderWithAppContext = (ui, { initialStates = new Array(), ...options } = {}) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, authUser);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]} {...props} />
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const renderWithAllContexts = (
  ui,
  { user = authUser, isDesktop = false, initialStates = new Array(), ...options } = {}
) => {
  const Wrapper = (props) => {
    return (
      <RecoilRoot
        initializeState={(snap) => {
          snap.set(userState, user);
          initialStates.forEach((s: { state: any; value: any }) => snap.set(s.state, s.value));
        }}
      >
        <ThemeProvider theme={themesLookup[defaultTheme]}>
          <ResponsiveContext.Provider value={{ width: isDesktop ? desktop : mobile }} {...props} />
        </ThemeProvider>
      </RecoilRoot>
    );
  };
  return testingLibraryRender(ui, { wrapper: Wrapper, ...options });
};

const render = renderWithAllContexts;
// Pass a different user
// const { getByTestId } = renderWithUserContext(<Dashboard />, { user: authUser });

export { renderWithUserContext, renderWithAppContext, renderWithAllContexts, render };
