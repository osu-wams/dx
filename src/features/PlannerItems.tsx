import React from 'react';
import Skeleton from 'react-loading-skeleton';
import { faFileEdit } from '@fortawesome/pro-light-svg-icons';
import { Card, CardHeader, CardContent, CardFooter, CardIcon } from '../ui/Card';
import { List } from '../ui/List';
import { useCourseSchedule } from '@osu-wams/hooks';
import { AuthorizeCanvas } from '../features/canvas/AuthorizeCanvas';
import Url from '../util/externalUrls.data';
import { ExternalLink } from '../ui/Link';
import { Event } from '../util/gaTracking';
import assignment from '../assets/assignment.svg';
import { EmptyState, EmptyStateImage, EmptyStateText } from '../ui/EmptyStates';
import { CanvasPlannerItems } from 'src/features/canvas/CanvasPlannerItems';
import { userState, plannerItemState } from 'src/state/application';
import { useRecoilValue } from 'recoil';

/**
 * Upcoming Assignments Card
 *
 * Displays upcoming assignments from Canvas.
 */
const PlannerItems = () => {
  const user = useRecoilValue(userState);
  const { data, isLoading } = useRecoilValue(plannerItemState);
  const courses = useCourseSchedule();

  const listOrEmpty = () => {
    if (isLoading) {
      return <Skeleton count={5} />;
    }

    if (!courses.loading && data && data.length && user.isCanvasOptIn === true) {
      return (
        <List>
          <CanvasPlannerItems data={data} courses={courses.data} />
        </List>
      );
    } else if (user.isCanvasOptIn === true) {
      return (
        <EmptyState>
          <EmptyStateImage src={assignment} alt="" />
          <EmptyStateText>You have no upcoming Canvas assignments</EmptyStateText>
        </EmptyState>
      );
    }
  };

  return (
    <Card>
      <CardHeader
        title="Canvas"
        badge={<CardIcon icon={faFileEdit} count={user.isCanvasOptIn ? data?.length : undefined} />}
      />
      <CardContent>
        {/* If not authorized to canvas, we display the link to have them authorize */}
        {!user.isCanvasOptIn && user.isCanvasOptIn !== undefined && <AuthorizeCanvas />}
        {listOrEmpty()}
      </CardContent>
      <CardFooter infoButtonId="canvas">
        {user.isCanvasOptIn === true && (
          <ExternalLink
            href={Url.canvas.main}
            onClick={() => Event('planner-items', 'View all link clicked')}
          >
            View more in Canvas
          </ExternalLink>
        )}
      </CardFooter>
    </Card>
  );
};

export default PlannerItems;
