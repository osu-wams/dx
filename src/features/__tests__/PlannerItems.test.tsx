import React from 'react';
import { waitForElement } from '@testing-library/react';
import user from '@testing-library/user-event';
import { render, mockAppContext, authUser } from '../../util/test-utils';
import PlannerItems from '../PlannerItems';
import { mockGAEvent } from '../../setupTests';
import { Student } from '@osu-wams/hooks';

const mockPlannerItems = Student.PlannerItems.mockPlannerItems;
const mockUsePlannerItems = jest.fn();
const mockNoData = { data: [], loading: false, error: false };

jest.mock('@osu-wams/hooks', () => {
  return {
    ...jest.requireActual('@osu-wams/hooks'),
    usePlannerItems: () => mockUsePlannerItems()
  };
});

describe('<PlannerItems />', () => {
  it('should have a "Week 5 Lab Discussion" assignment on our mock data', async () => {
    mockUsePlannerItems.mockReturnValue(mockPlannerItems);
    const { findByText } = render(<PlannerItems />);
    await findByText('Week 5 Lab Discussion');
  });

  it('should track analytics when footer link and assignment is clicked', async () => {
    mockUsePlannerItems.mockReturnValue(mockPlannerItems);
    const { getByText, findByText } = render(<PlannerItems />);

    // Planner Item
    await waitForElement(() => findByText('Week 5 Lab Discussion'));

    const PlannerItem = getByText('Week 5 Lab Discussion');
    user.click(PlannerItem);
    expect(mockGAEvent).toHaveBeenCalled();

    // Footer link
    const CanvasLink = await findByText('View more in Canvas');
    user.click(CanvasLink);
    expect(mockGAEvent).toHaveBeenCalled();
  });

  it('should find empty state if our promise returns empty', async () => {
    mockUsePlannerItems.mockReturnValue(mockNoData);
    const { findByText } = render(<PlannerItems />);
    const noAssignments = await findByText(/You have no upcoming Canvas assignments/);
    expect(noAssignments).toBeInTheDocument();
  });
});

describe('with an InfoButton in the CardFooter', () => {
  // Set mock function result before running any tests
  beforeEach(() => {
    mockUsePlannerItems.mockReturnValueOnce(mockPlannerItems);
  });

  const validIinfoButtonId = 'canvas';

  it('does not display the button when the infoButtonData is missing it', async () => {
    const testAppContext = {
      ...mockAppContext,
      infoButtonData: [{ id: 'invalid-id', content: 'content', title: 'title' }]
    };
    const { queryByTestId, findByText } = render(<PlannerItems />, {
      appContext: testAppContext
    });
    const noAssignments = await findByText(/You have no upcoming Canvas assignments/);
    expect(noAssignments).toBeInTheDocument();

    const element = queryByTestId(validIinfoButtonId);
    expect(element).not.toBeInTheDocument();
  });

  it('displays the button when the infoButtonData is included', async () => {
    const testAppContext = {
      ...mockAppContext,
      infoButtonData: [{ id: validIinfoButtonId, content: 'content', title: 'title' }]
    };
    const { findByTestId } = render(<PlannerItems />, { appContext: testAppContext });

    const element = await findByTestId(validIinfoButtonId);
    expect(element).toBeInTheDocument();
  });
});

describe('with a user who has not opted-in Canvas', () => {
  it('hides the badge count and shows the authorization call to action', async () => {
    mockUsePlannerItems.mockReturnValueOnce(mockPlannerItems);
    const mockUser = authUser;
    mockUser.data.isCanvasOptIn = false;
    mockUser.isCanvasOptIn = false;

    const { queryByTestId, findAllByText } = render(<PlannerItems />, { user: mockUser });

    const element = queryByTestId('icon-counter');
    expect(element).not.toBeInTheDocument();
    const authorizeElements = await findAllByText(/Authorize Canvas/i);
    expect(authorizeElements.length).toBe(2);
  });
});
