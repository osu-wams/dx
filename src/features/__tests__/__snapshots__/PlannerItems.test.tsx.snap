// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlannerItems /> should have a "My Awesome Planner Note" note on our mock data: without hover 1`] = `
.c0 {
  list-style-type: none;
}

.c0:last-child {
  margin-bottom: 16px;
}

.c0 > button {
  cursor: pointer;
}

.c5 {
  padding-right: 16px;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  text-align: left;
}

.c6 {
  color: #423E3C;
  margin: 0;
  font-weight: normal;
}

.c2 {
  width: 100%;
  background: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  border-radius: 8px;
  -webkit-transition: all 150ms ease-in-out 0s;
  transition: all 150ms ease-in-out 0s;
  padding: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 svg,
.c2 img {
  height: 3rem;
  font-size: 2.4rem;
  width: 3rem !important;
}

.c2 svg + .c4,
.c2 img + .c4 {
  padding-left: 1.5rem;
}

.c2 .warning-icon {
  font-size: 14px;
  width: 14px !important;
  margin-top: -10px;
  margin-left: 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
}

.c1 .MuiCheckbox-root {
  width: 48px;
  height: 48px;
  font-size: 18px;
  margin-right: 5px;
}

.c1 .c4 {
  padding-right: 0;
}

.c7 {
  color: #7B746F;
  font-size: 14px;
  line-height: 1.6rem;
}

.c3 {
  line-height: 14px;
  text-align: center;
  padding: 0 16px 0 0;
  color: #D73F09;
  width: 58px;
}

.c3 strong {
  line-height: 20px;
  font-size: 20px;
}

.c8 {
  background: #E9E5E4;
  border-radius: 4px;
  color: #2E2B2A;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin-top: 4px;
  text-align: center;
  white-space: nowrap;
}

@media (min-width:768px) {

}

<li
  class="c0 c1"
>
  <div
    class="c2"
  >
    <div
      class="c3"
      data-testid="course-list-item-header"
    >
      <div>
        WR
      </div>
      <div>
        <strong>
          214
        </strong>
      </div>
    </div>
    <div
      class="c4 c5"
    >
      <h4
        class="c6"
      >
        My Awesome Planner Note
      </h4>
      <div
        class="c7"
      >
        Due Jan 1st at 1:01 AM
      </div>
    </div>
    <span
      class="c8"
    >
      discussion topic
    </span>
  </div>
</li>
`;

exports[`<PlannerItems /> should have a "Week 5 Lab Discussion" assignment on our mock data: with hover 1`] = `
.c5 {
  color: #D73F09;
  background-color: transparent;
}

.c0 {
  list-style-type: none;
}

.c0:last-child {
  margin-bottom: 16px;
}

.c0 > button {
  cursor: pointer;
}

.c7 {
  padding-right: 16px;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  text-align: left;
}

.c9 {
  color: #423E3C;
  margin: 0;
  font-weight: normal;
}

.c3 {
  width: 100%;
  background: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  border-radius: 8px;
  -webkit-transition: all 150ms ease-in-out 0s;
  transition: all 150ms ease-in-out 0s;
  padding: 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3 svg,
.c3 img {
  height: 3rem;
  font-size: 2.4rem;
  width: 3rem !important;
}

.c3 svg + .c6,
.c3 img + .c6 {
  padding-left: 1.5rem;
}

.c3 .warning-icon {
  font-size: 14px;
  width: 14px !important;
  margin-top: -10px;
  margin-left: 5px;
}

.c4 {
  color: #423E3C;
  cursor: pointer;
}

.c4:hover {
  border: 1px solid transparent;
}

.c4:hover .c8,
.c4:hover .List__ListItemContentLinkName-sc-1h14hxl-4 {
  color: #D73F09;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
}

.c1 .MuiCheckbox-root {
  width: 48px;
  height: 48px;
  font-size: 18px;
  margin-right: 5px;
}

.c1 .c6 {
  padding-right: 0;
}

.c1:hover {
  box-shadow: rgba(66,62,60,0.1) 0px 10px 16px,rgba(105,99,97,0.05) 0px 3px 16px;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  border-color: transparent;
}

.c1:hover .c8,
.c1:hover .c2 {
  color: #D73F09;
}

.c10 {
  color: #7B746F;
  font-size: 14px;
  line-height: 1.6rem;
}

.c11 {
  background: #E9E5E4;
  border-radius: 4px;
  color: #2E2B2A;
  font-size: 12px;
  padding: 2px 4px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin-top: 4px;
  text-align: center;
  white-space: nowrap;
}

@media (min-width:768px) {

}

<li
  class="c0 c1"
>
  <a
    class="c2 c3 c4"
    href="https://canvas.oregonstate.edu/courses/1724943/assignments/7547409/submissions/6227907"
    target="_blank"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-file-edit fa-w-12 c5"
      color="#D73F09"
      data-icon="file-edit"
      data-prefix="fal"
      data-testid="planner-icon"
      focusable="false"
      role="img"
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M369.9 97.9L286 14C277 5 264.8-.1 252.1-.1H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48V131.9c0-12.7-5.1-25-14.1-34zm-22.6 22.7c2.1 2.1 3.5 4.6 4.2 7.4H256V32.5c2.8.7 5.3 2.1 7.4 4.2l83.9 83.9zM336 480H48c-8.8 0-16-7.2-16-16V48c0-8.8 7.2-16 16-16h176v104c0 13.3 10.7 24 24 24h104v304c0 8.8-7.2 16-16 16zM219.2 247.2l29.6 29.6c1.8 1.8 1.8 4.6 0 6.4L136.4 395.6l-30.1 4.3c-5.9.8-11-4.2-10.2-10.2l4.3-30.1 112.4-112.4c1.8-1.8 4.6-1.8 6.4 0zm64.4 1.2l-16.4 16.4c-1.8 1.8-4.6 1.8-6.4 0l-29.6-29.6c-1.8-1.8-1.8-4.6 0-6.4l16.4-16.4c5.9-5.9 15.4-5.9 21.2 0l14.8 14.8c5.9 5.8 5.9 15.3 0 21.2z"
        fill="currentColor"
      />
    </svg>
    <div
      class="c6 c7"
    >
      <h4
        class="c8 c9"
      >
        Week 5 Lab Discussion
      </h4>
      <div
        class="c10"
      >
        Due May 2nd at 6:59 AM
      </div>
    </div>
    <span
      class="c11"
    >
      discussion topic
    </span>
  </a>
</li>
`;
